### PR TITLE
add miscellaneous notes to ecr viewer

### LINKED ISSUE #1073: Add "Clinical Notes" sub-section
## Action Requested
Add the "Clinical notes" sub-section ot the Clinical Info section of the eCR Viewer.

## Acceptance Criteria
The following fields are included in the eCR Viewer:

* Miscellaneous Notes

## Testing
Unit testing will be required.

## Background & Context
Miscellaneous (doctor's) notes are one of the most common fields case investigators look for in a jurisdiction.

## Designs
Screenshot 2024-01-12 at 2 28 52 PM

## Additional Information
We need to validate that this section is being correctly converted to FHIR. Miscellaneous notes appear to be converted by default in the MS FHIR Converter, but it's been tricky to find eCR records that actually have the notes included to verify. If it is present, it will probably be in the "Miscellaneous Notes" section of the FHIR bundle, with entries listed at the end of the section. 
Sample: https://github.com/CDCgov/phdi/blob/2ce3f5cd959378179db0873c662d7a8ee89ce25a/containers/ecr-viewer/seed-scripts/fhir_data/8675309a-7754-r2d2-c3p0-973d9f777777.json#L55

### PR DESCRIPTION
**Pull request by BobanL**
# PULL REQUEST

## Summary
- ECR Viewer
  - Add Clinical notes to the Clinical Info section
  - Install [`html-react-parser` to render the HTML](https://github.com/remarkablemark/html-react-parser) and [classnames](https://github.com/JedWatson/classnames) 
  - Add optional classNames variables for `DataDisplay` 
  - Regenerate all fhir_data with the convert-seed-data script. 
  - Minor: Add types to `returnImmunizations` parameters. 
- Fhir Converter
  - Add new method for `ToHtmlString` to handle converting the XML/dictionary back into HTML text. 
    - Handles `list`, `dictionary<string, object>`, and `string` values based on the examples from LAC.  
    - Only a limited number of tags are supported right now, but can be expanded with the  `supportedTags` variable
      - `"br", "li", "ol", "p", "span", "table", "tbody", "td", "textarea", "th", "thead", "tr", "u", "ul"` 
    - Converts certain XML tags to html like `List` to `ul` and `item` to `li` 
  - Create tests for `CustomFilters.cs` - this can be ran with `dotnet test` or `npm run test` :) 
## Related Issue
Fixes #1073 

## Additional Information
Example screenshots:
image
image

## Checklist

- [ ] If this code affects the other scrum team, have they been notified? (In Slack, as reviewers, etc.)

[//]: # (PR title: Remember to name your PR descriptively!)

### CODE REVIEW

**Comment by sarahtress** [comment]
> @sarahtress based on the screenshots, and the limited data i have for miscellanous notes here are two examples of what it could look like. The first example of Narrative text is a table and just has the default blank styling. Do you want me to match the styling that is in the other tables in this section? image image

@BobanL thanks for sending! I guess if there is a table within it, then it makes sense to match the table styling we have for the rest of the page. for the first screenshot you sent - can we make sure there isn't any bolded text within the notes themselves? would that be feasible? it makes the hierarchy look visually confusing

**Comment by sarahtress** [comment]
> > > @sarahtress based on the screenshots, and the limited data i have for miscellanous notes here are two examples of what it could look like. The first example of Narrative text is a table and just has the default blank styling. Do you want me to match the styling that is in the other tables in this section? image image
> > 
> > 
> > @BobanL thanks for sending! I guess if there is a table within it, then it makes sense to match the table styling we have for the rest of the page. for the first screenshot you sent - can we make sure there isn't any bolded text within the notes themselves (i.e. can "Narrative Text" be made not bold)? would that be feasible? it makes the hierarchy look visually confusing
> 
> So we met about this, but for anyone else looking at this Narrative text is a table actually 😆
> 
> @sarahtress With the formatting properly applied it looks like this below. And there are also times where it has lists in it, and this is what it looks like. Happy to get on a call to discuss if it is easier than leaving a message. image image

@BobanL this looks good to me for now. might want to revisit in the future if we see more funky formatting or if the data ends up being extremely long, but for now let's keep this! thanks

File: containers/fhir-converter/CustomFhir.sln
Comment (nickbristow) [inline containers/fhir-converter/CustomFhir.sln:1]: what is this??

File: containers/fhir-converter/CustomFhir/CustomFiltersTests.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        var strList = new List<object>() { "Race", "car" };
+        var dictDict = new Dictionary<string, object>(){
+            {"span", strList},
+            {"/nSun", "flower"}
+        };
+        var actual = Filters.ToHtmlString(dictDict);
+        Assert.Equal("<span>Race</span><span>car</span>", actual);
+    }
+
+    // from 020e83d2-a1ca-4056-bc9f-6cffe8609e99 Active Problems
```
Comment (nickbristow) [inline containers/fhir-converter/CustomFhir/CustomFiltersTests.cs:71]: do we need this comment?

**Review by nickbristow — APPROVED** [review]

File: containers/fhir-converter/CustomFhir.sln
Comment (BobanL) [reply containers/fhir-converter/CustomFhir.sln:1]: This is a [solution file for dotnet](https://docs.fileformat.com/programming/sln/#:~:text=SLN%20extension%20represents%20a%20Visual,file%20in%20any%20text%20editor.). This was autogenerated when I went through the create new project setups steps in vscode and if you delete this the tests will not run.

Comment (nickbristow) [reply containers/fhir-converter/CustomFhir.sln:1]: 👍

**Review by emmastephenson — APPROVED** [review]
Looks great, thank you for making these changes!

The only thing I think is missing is an update to the `test` workflow to make sure the dotnet tests are actually run during CI. But if you want to save that for a followup PR, feel free - this PR is pretty long already.

File: containers/fhir-converter/CustomFhir/CustomFilters.cs [hunk 2]
```diff
@@ -76,5 +82,73 @@ public static string ConcatenateTds(IDictionary<string, object> data)
...
+    private static string CleanStringFromTabs(string value)
+    {
+        const string reduceMultiSpace = @"[ ]{2,}";
+        return Regex.Replace(value.Replace("\t"," "), reduceMultiSpace, " ");
+    }
+
+    public static string ToHtmlString(object data)
+    {
+       var stringBuilder = new StringBuilder();
+       if(data is string stringData)
```
[the remark this answers is not in the document]
Comment (BobanL) [reply containers/fhir-converter/CustomFhir/CustomFilters.cs:115]: [The `is` check is a type check and the last part is casting the value to a variable. ](https://www.dotnetperls.com/is#:~:text=Is%2C%20pattern%20matching.%20Here%20we%20see%20a%20separate%20syntax%20form%20for%20the%20%22is%22%20operator.%20We%20have%20an%20object%2C%20and%20we%20want%20to%20cast%20it%20to%20a%20string.%20We%20use%20pattern%20matching%20for%20this.) 
It is equivalent to doing 
```dotnet
if(data is string)
{
    string stringData = (String) data;
}
```

**Pull request merged**